llvm/lib/SYCLLowerIR/ModuleSplitter.cpp

```diff
@@ -667,11 +667,13 @@ void ModuleDesc::restoreLinkageOfDirectInvokeSimdTargets() {
 // the transformation safe.
 static bool mustPreserveGV(const GlobalValue &GV) {
   if (const Function *F = dyn_cast<Function>(&GV)) {
-    // When dynamic linking is supported, we internalize everything that can
-    // not be imported which also means that there is no point of having it
+    // When dynamic linking is supported, we internalize everything (except
+    // kernels which are the entry points from host code to device code) that
+    // cannot be imported which also means that there is no point of having it
     // visible outside of the current module.
     if (SupportDynamicLinking)
-      return canBeImportedFunction(*F);
+      return canBeImportedFunction(*F) ||
+             F->getCallingConv() == CallingConv::SPIR_KERNEL;
```

Comment:
I would swap the two checks, because calling convention check is cheaper and therefore we will be able to avoid calling more expensive `canBeImportedFunction` for some functions in a module

Reply:
Done.